README
NAME
    Regexp::Optimizer - optimizes regular expressions

SYNOPSIS
      use Regexp::Optimizer;
      my $o  = Regexp::List->new;
      my $re = $o->optimize(qr/foobar|fooxar|foozap/);
      # $re is now qr/foo(?:[bx]ar|zap)/

ABSTRACT
    This module does, ahem, attempts to, optimize regular expressions.

INSTALLATION
    To install this module type the following:

       perl Makefile.PL
       make
       make test
       make install

DESCRIPTION
    Here is a quote from perltodo

      Factoring out common suffices/prefices in regexps (trie optimization)

           Currently, the user has to optimize "foo|far" and "foo|goo" into
           "f(?:oo|ar)" and "[fg]oo" by hand; this could be done automatically.

    This module implements just that.

  EXPORT
    Since this is an OO module there is no symbol exported.

METHODS
    This module is implemented as a subclass of Regexp::List. For methods
    not listed here, see Regexp::List.

    $o = Regexp::Optimizer->new;
    $re = $o->optimize(*regexp*);
        Does the job. Note that unlike "->list2re()" in Regexp::List, the
        argument is the regular expression itself. What it basically does is
        to find groups will alterations and replace it with the result of
        "$o->list2re".

    $re = $o->list2re(*list of words ...*)
        Same as "list2re()" in Regexp::List in terms of functionality but
        how it tokenize "atoms" is different since the arguments can be
        regular expressions, not just strings. Here is a brief example.

          my @expr = qw/foobar fooba+/;
          Regexp::List->new->list2re(@expr) eq qr/fooba[\+r]/;
          Regexp::Optimizer->new->list2re(@expr) eq qr/foob(?:a+ar)/;

CAVEATS
    This module is still experimental. Do not assume that the result is the
    same as the unoptimized version.

    *   When you just want a regular expression which matches normal words
        with not metacharacters, use <Regexp::List>. It's more robus and
        much faster.

    *   When you have a list of regular expessions which you want to
        aggregate, use "list2re" of THIS MODULE.

    *   Use "->optimize()" when and only when you already have a big regular
        expression with alterations therein.

        "->optimize()" does support nested groups but its parser is not
        tested very well.

BUGS
    *   Regex parser in this module (which itself is implemented by regular
        expression) is not as thoroughly tested as Regex::List

    *   Does not grok (?{expression}) and (?(cond)yes|no) constructs yet

SEE ALSO
    Regexp::List -- upon which this module is based

    "eg/" directory in this package contains example scripts.

    Perl standard documents
         L<perltodo>, L<perlre>

    CPAN Modules
        Regexp::Presuf, Text::Trie

    Books
        Mastering Regular Expressions
        <http://www.oreilly.com/catalog/regex2/>

AUTHOR
    Dan Kogai <dankogai@dan.co.jp>

COPYRIGHT AND LICENSE
    Copyright 2003 by Dan Kogai

    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
#
# $Id: Changes,v 0.15 2004/12/05 16:07:34 dankogai Exp dankogai $
#

$Revision: 0.15 $ $Date: 2004/12/05 16:07:34 $
! lib/Regexp/Optimizer.pm lib/Regexp/List.pm
  Pod fixed accordingly to RT # 8733
  http://rt.cpan.org/NoAuth/Bug.html?id=8733

0.14 2004/11/05 12:44:48
! t/02-list.t
  Addressed test failure that was raised by Perl 5.8.5
  http://rt.cpan.org/NoAuth/Bug.html?id=8165

0.13 2004/05/08 05:55:35
! lib/Regexp/Optimizer.pm
  Document bug, reported by Frederick in the mail below, fixed.
! Makefile.PL
! Regexp/ -> lib/Regexp/
  Module hierarchy realigned so cygwin is happy.  
  Reported by Frederick Weiland
  <30F579C44E1598429167C5E20E307BA19D807D@atlmsg01.raremedium.net>

0.12 2004/05/04 17:12:14
! Regexp/Optimizer.pm
  Perl 5.8.4 and later corrected a bug so it is fatal to go
    my $x = qr{ ... (??{ $x }) ... };
  under "use strict".  Unfortunately Regexp/Optimizer.pm had two
  occurances thereof.  Now fixed.

0.11 2004/05/03 15:09:14
! Regexp/List.pm
  Deep recursion addressed and corrected by yoz@yoz.com
  https://rt.cpan.org/Ticket/Display.html?id=4937

0.10 2003/06/02 20:11:07
! *
  Version jumps to 0.10 do to radical changes, especially in
  $Regexp::Optimize.
! Regexp/Optimizer.pm t/02-list.t t/03-utf8.t
  Aagh!  Why didn't I come up with such a simple idea !?
  ->optimize is completely rewritten.
  t/*.t is streamlined to reflect changes.
  Now ->optimize() *really* optimizes even for
  qr/(?:1|12)|123/.  Used to return qr/1(?:2?|23)/ but now it
  returns qr/1(?:23?)?/o
! Regexp/Optimizer.pm  Regexp/List.pm
  POD enhanced.

0.02 2003/06/01 00:11:26
! Regexp/Optimizer.pm  Regexp/List.pm
  * Lots of bug fixes regarding $o->optimize() for nested parens
! t/02-list.t
  Test data for #21 corrected
! t/03-utf8.t
  buggy SKIP: sections fixed.

0.01 2003/05/31 10:44:41
+ *
  0th release


t/00-dict.pl
#
# $Id: 00-dict.pl,v 0.1 2003/05/31 10:43:20 dankogai Exp $
#
use strict;
# use Test::More qw/no_plan/;
use Regexp::List;
use Getopt::Std;
use File::Basename;
use File::Spec;
use Time::HiRes qw/time/;

my $dict_file   = shift or die;
my $regexp_file = File::Spec->catfile('t', basename($dict_file) . '.rx');
my @words;
my ($now, $then, $total);

sub comment{
    my ($msg, $check) = @_;
    $now = time(); 
    $check and $then = $now;
    printf "# %s (%13.6f s) $msg\n", scalar(localtime($now)), $now - $then;
}

unless (-f $regexp_file){
    comment "Reading $dict_file", 1;
    open DICT, $dict_file or die "$dict_file:$!";
    my $nlines = 0;
    while(<DICT>){
	chomp;
	push @words, $_;
	$nlines++;
    }
    comment "$nlines lines read.";
    comment "Making regexp.", 1;
    my $l = Regexp::List->new;
    my $regexp = $l->list2regex(@words);
    comment "Done.";
    comment "Saving to $regexp_file", 1;
    open RX, ">", $regexp_file or die "$regexp_file:$!";
    print RX 'qr/' . $regexp . '/';
    close RX;
    undef $regexp;
}

comment "Reading $regexp_file", 1;
my $regexp = do $regexp_file;
comment ($@ ? "FAILED: $@" : "Done.");

comment "Opening $dict_file for comparison.", 1;
open DICT, $dict_file or die "$dict_file:$!";
my $nlines = 0;
$nlines++ while(<DICT>);
comment "$dict_file:$nlines lines found.";
seek DICT, 0, 0;
comment "Showtime!", 1;
my $line = 0; $| = 1;
while(<DICT>){
    m/($regexp)/;
    chomp;
    $_ eq $1 or die "$_ did not match! (word = $_, match = $1)";
    printf "\r# %d/%d", ++$line, $nlines;
}
print "\n";
comment "Done.";
comment sprintf("%.3f matches/s", $nlines/($now-$then))
		


eg/mkregexp.pl
#!/usr/local/bin/perl
#
# $Id: mkregexp.pl,v 0.1 2003/05/31 10:43:20 dankogai Exp $
#
use strict;
use Regexp::List;
use Getopt::Std;

my %Opt;
getopts("a:dixw:" => \%Opt);
my $re_word = 
    $Opt{w} ? qr/$Opt{w}/ : qr/[0-9A-Za-z_\x80-\xff]+/o;
my $sub = '';
if ($Opt{a}){
    $sub = eval qq{ sub { $Opt{a} } };
    $@ and die "$Opt{a} : $@";
}

my %words;

unless ($Opt{a}){
    while(<>){
	s/($re_word)/$words{$1}++/ego;
    }
}else{
    for my $file (@ARGV){
	open F, $file or die "$file:$!";
	while(<F>){
	    s/($re_word)/$words{$1}++/ego;
	}
    }
}
#use Data::Dumper;
#warn join "\n",  sort keys %words;

my $modifiers;
$Opt{i} and $modifiers .= 'i';
$Opt{x} and $modifiers .= 'x';

my $l2r = Regexp::List->new(
			   debug     =>  $Opt{d},
			   as_string => !$Opt{a},
			   # sort => 1,
			   optim_sx => 1,
			   optim_px => 1,
			   modifiers =>  $modifiers,
			  );
my $regex = $l2r->list2regex(keys %words);
unless ($Opt{a}){
    print "qr/$regex/\n";
}else{
    for my $file (@ARGV){
	open F, $file or die "$file:$!";
	while(<F>){
	    s/($regex)/$sub->($1)/ego;
	    print; 
	}
    }
}


lib/Regexp/List.pm
#
# $Id: List.pm,v 0.13 2004/12/05 16:07:34 dankogai Exp dankogai $
#
package Regexp::List;
use 5.006; # qr/(??{}/ needed
use strict;
use warnings;
no warnings 'uninitialized';
#use base qw/Exporter/;
our $VERSION = '0.16_001';

#our @EXPORT = qw();
#our %EXPORT_TAGS = ( 'all' => [ qw() ] );
#our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );
our $DEBUG     = 0;

our $FILLER = "\x{fffd}"; # fallback

# According "perlre", a "^" immediately after the "?" is a shorthand
# equivalent to "d-imsx" since Perl 5.14.
my $options = 'imsx\-';
if ($^V gt v5.14.0) {
    $options .= '^';
}

our $RE_PAREN; # predeclear
our $RE_START =
    qr{(?:
    (?!\\)\((?:\?
    (?:
     ([$options]*:)| # options 
     \<?[\=\!]     | # look(behind|ahead)
     \#[^\)]+      | # comments
     #$RE_PAREN    | # ( condtion )
     #\??$RE_EXPR  | # { expression }
     \>              # independent subexpression
    ))?
    )}xo;

our $RE_XCHAR =
    qr{
       (?:\\
	(?:
	[^0xclupPNLUQEXC]     | # ordinary escaped character
	 0[0-9][0-9]          | # octal
	 x(?:[0-9A-Fa-f]{1,2} | # hex
	   \{[0-9A-Fa-f]+\})  | # unicode hex
	 c.                   | # control char
	 [pP]\{\w+\}          | # unicode properties
	 N\{[\w\ ]+\}         | # unicode name
	 )
	)}xo;

our $RE_PCHAR = 
    qr{
       (?:\\
        (?:
         [XC]  # unicode name
	)
       )}xo;

our $RE_CHAR = 
    qr{(?:
        (?!\\)[^\\]            | # raw character (except \)
        $RE_XCHAR              | # extended character
       )
      }xo;

our %PARAM = 
    (
     _i         => 0,
     _m         => 0,
     _s         => 0,
     _x         => 0,
     _char      => $RE_CHAR,
     _token     => $RE_CHAR,
     _cclass    => $RE_CHAR,
     debug      => $DEBUG,
     capture    => 0,
     lookahead  => 1,
     modifiers  => '',
     optim_cc   => 1,
     optim_cq   => 1,
     optim_sx   => 1,
     po         => '(?:',
     pc         => ')',
     quotemeta  => 1, 
     sort       => 0,
     );

# aliases

*l2r         = \&list2re;
*list2regex  = \&list2re;
*list2regexp = \&list2re;

sub new{
    my $class = ref $_[0] ? ref shift : shift;
    my $self = bless { %PARAM } => $class;
    $self->set(@_);
}

sub clone{
    my $self = shift;
    my $clone = bless { %$self } => ref $self;
    $clone->set(@_);
}

sub set{
    my $self = shift;
    my %param = @_;
    for (sort keys %param){
	$self->{$_} = $param{$_};
	if ($_ eq 'capture'){
	    $self->{op} = $self->{capture} ? '(' : '(?:';
	    $self->{cp} = ')';
	}
	if ($_ eq 'modifiers'){
	    map { $self->{'_' . $_} = 0 } qw/i m s x/;
	    map { $self->{'_' . $_} = 1 } split //, $self->{$_};
	}
    }
    $self;
}

sub tokens{
    my $self  = shift;
    my $str   = shift;
    grep {$_ ne '' } split /($self->{_token})/, $str;
}

sub regopt{
    my  $re = shift;
    #ref $re eq 'Regexp' or return;
    $re =~ /^($RE_START)/ or return; # die "malformed regexp : $re";
    my $opt = $1;
    $opt =~ s/\(\?//o; $opt =~ s/[-:].*//o;
    $opt;
}

sub expand{
    my $self  = shift;
    my $re   = shift;
    my $isre  = ref $re eq 'Regexp';
    #$isre or $re = qr/$re/;
    my $mod = regopt($re);
    $mod =~ /x/ or $mod .= 'x';
    my ($indent, @indent);
    $re =~ 
	s{
	  ( $RE_START | (?!\\)[\)|])
	 }{
	     my $paren = $1;
	     my $sub = $paren;
	     if  ($paren eq ')'){ # close
		 $indent -= pop @indent;
	     }elsif($paren eq '|'){ # |
		 $sub = " | \n";
		 $sub .= " " x $indent;
	     }else{
		 $sub  = $indent ? "\n" : '';
		 $sub .= " " x $indent . $paren;
		 $indent += length($paren);
		 push @indent, length($paren);
	     }
	     $sub;
	 }geox;
    $isre ? qr/(?$mod:$re)/ : qq/(?$mod:$re)/;
}

sub unexpand{
    my $self = shift;
    my $re   = shift;
    my $isre  = ref $re eq 'Regexp';
    my $mod = regopt($re); 
    $mod =~ s/x//o;
    $re =~ s/\((?!\\)\?\#[^\)]+\)//o; # strip (?#comment)
    $re =~ s/(?!\\)#.*$//mg;          # strip comment
    $re =~ s/(?!\\)[ \t]//g;          # strip space
    # $re =~ s/([^\x00-\xff])/sprintf('\x{%04x}', ord($1))/eg;
    # and finally strip CRLF
    $re =~ s/[\n\r]//g;
    $isre ? 
	$mod ? qr/(?$mod:$re)/ : qr/$re/ :
	$mod ? qq/(?$mod:$re)/ : $re;
}

sub list2re { 
    use utf8; # for substr
    no warnings 'redefine'; # for cheats
    my $self = shift;
    # trie construction allows no duplicates 
    # so we make sure they are all unique
    my (%list, @list);
    # Unique with order preserved
    if ($self->{_i}){ for (@_){ $_=lc($_); $list{$_}++ or push @list, $_ } }
    else            { for (@_){            $list{$_}++ or push @list, $_ } }
    undef %list; # to save memory
    #$self->{sort} and @list = sort {length($b) <=> length($a) } @list;
    $self->{sort} and @list = sort @list;
    my $result;
    if ($self->{quotemeta}){
	# cheat
	*_head = sub{ substr($_[1], 0, $_[2]*2) };
	*_tail = sub{ substr($_[1], $_[2]*2) };
	$result = _trie_regex($self, map { _metaquote($_) } @list);
	$result =~ tr/\x00//d; 
	#$result =~ tr/\x{FFFd}//d;
    }else{
	*_head = \&_head_re;
	*_tail = \&_tail_re;
	$self->{_x} and @list = map { s/\\? /\\ /g; $_  } @list;
	$result = _trie_regex($self, @list); 
    }
    my $lookahead;
    if ($self->{lookahead}){
	my %lookahead;
	$lookahead{$self->_first($_)}++ for @list;
        my @lookahead = 
	    $self->{quotemeta} 
		?  map { tr/\x00//d; qq/\Q$_/ } keys %lookahead
		    #map { tr/\x{FFFd}//d; qq/\Q$_/ } keys %lookahead
		    : keys %lookahead;
	@lookahead = sort sort grep {!/^\\[luLUEQXC]/} @lookahead;
	if (@lookahead > 1){
	    my $lookahead = join('' => @lookahead);
	    $result = qq/(?=[$lookahead])$result/;
	}
	undef @lookahead;
    }
    my $mod = $self->{modifiers};
    $mod =~ 'x' and return $self->expand($result);
    $result = $self->{as_string} ? 
	$mod ? qq/(?$mod:$result)/ : qq/$result/ :
	    $mod ? qr/(?$mod:$result)/ : qr/$result/;
}

sub _metaquote{
    my $str =
	join '' => 
	    map { my $q=qq/\Q$_/; length($q) == 2 ? $q : "\x00$q" }
	    #map { my $q=qq/\Q$_/; length($q) == 2 ? $q : "\x{FFFd}$q" }
		split // => shift;
    $str;
}

sub _first{
    my $self = shift;
    my $str = shift;
    my $re = $self->{_char};
    $str =~ /^($re)/o;
    return $1;
}

sub _head_re{
    my $self = shift;
    my ($str, $pos) = @_;
    $str eq '' and return '';
    my $token = $self->{_token};
    for (my $p = $pos, pos($str) = 0; $p > 0 ; $p--){
	$str =~ /\G$token/gc or last;
    }
    substr($str, 0, pos($str));
    
}

sub _tail_re{
    use utf8;
    my $self = shift;
    my ($str, $pos) = @_;
    $str eq '' and return '';
    my $token = $self->{_token};
    for (my $p = $pos, pos($str) = 0; $p > 0 ; $p--){
	$str =~ /\G$token/gcs or last;
    }
    substr($str,pos($str));
}

use Data::Dumper;
$Data::Dumper::Indent = 1;

sub _prefixes {
    my $self = shift;
    my (@head, @prefix, %prefix);
    for (@_) {
	my $c = $self->_head($_, 1);
	exists $prefix{$c} or push @prefix, $c; # to preserve order
	$prefix{$c} ||= [];
	push @{$prefix{$c}}, $_;
    }
    for (@prefix) {
	# Find common substring
	my $prefix = $prefix{$_}->[0];
	if (@{$prefix{$_}} == 1){
	    push @head, [ $prefix ]; next 
	}
	my $l = length($prefix);
	for (@{$prefix{$_}}) {
	    $l -= 1
		while $self->_head($_, $l) ne $self->_head($prefix, $l);
	}
	# Return value
	$prefix = $self->_head($prefix, $l);
	my @suffix = map {$self->_tail($_, $l)} @{$prefix{$_}};
	push @head, [$prefix, @suffix];
    }
    #print Dumper \@head;
    #sleep 1;
    @head;
}


sub _rev{
    my $self = shift;
    my $str = shift;
    if ($self->{quotemeta}){
	return length $str > 2 ?
	    join '' => reverse split /(..)/, $str : $str;
    }else{
	my $re = $self->{_token};
	#return $str =~ /^$re?$/o ?
	#   $str : join '' => reverse split /($re)/, $str;
	$str =~ /^$self->{_token}$/ and return $str;
	my @token;
	$str =~ s{ ($re) }{ push @token, $1 }egx;
	return join '' => reverse @token;
	    
	    
    }
}
sub _trie_regex {
    my $self = shift;
    @_ or return;
    @_ == 1 and return shift;

    $self->{debug} and $self->{_indent}++;
    $self->{debug} and
	print STDERR '>'x $self->{_indent}, " ", join(',' => @_), "\n";

    my (@leaf, @result);

    #
    # Suffixing Optimization
    # - only leaf nodes in the same branch can be suffix-bundled
    #
    if ($self->{optim_sx}){
	for ($self->_prefixes(@_)){
	    my ($prefix, @suffix) = @$_;
	    if (@suffix){
		push @result, $prefix.$self->_trie_regex(@suffix);
	    }else{
		push @leaf, $prefix;
	    }
	}
	for ($self->_prefixes(map { $self->_rev($_) } @leaf)){
	    my ($suffix, @prefix) = @$_;
	    $suffix = $self->_rev($suffix);
	    if (@prefix){
		push @result, 
		    $self->_trie_regex(map { $self->_rev($_) } @prefix)
			. $suffix;
	    }else{
		push @result, $suffix;
	    }
	}
    }else{
	for ($self->_prefixes(@_)){
	    my ($prefix, @suffix) = @$_;
	    push @result, @suffix ? $prefix.$self->_trie_regex(@suffix) : $prefix;
	}
    }

    my $result;

  RESULT:
    {
	@result == 1 and $result = $result[0] and last RESULT;
	my $q = '';
	# alteration check
	# we do linear seach here to preserve order.
	for (my $i = 0; $i < @result; $i++){
	    if ($result[$i] eq ''){
		 splice @result, $i, 1;
		 $q = '?';
		 last;
	    }
	}
	# if ($result[0] eq '') { $q = '?'; shift @result }
	# extract character class
	if ($self->{optim_cc}){
	    my @char; my  $charpos = -1;
	    for (my $i = 0; $i < @result; $i++){
		if ($self->{quotemeta}){
		    if (length($result[$i]) == 2){ 
			$charpos < 0 and $charpos = $i;
			push @char => splice(@result, $i, 1, "");
		    }
		}else{
		    if ($result[$i] =~ /^$self->{_cclass}$/){
			$charpos < 0 and $charpos = $i;
			push @char => splice(@result, $i, 1, "");
		    }
		}
	    }
	    if (@char){
		my $char = $self->_optim_cc(@char);
		splice @result, $charpos, 0, $char;
		@result = grep {$_} @result;
		if (@result == 1){
		    $result = "$result[0]$q" and last RESULT;
		}
	    }
	}
	my $joiner = '|' ;
	if ($self->{optim_cq} and @result == 1 and
	    ($self->{quotemeta} 
	     ? length($result[0]) ==
	     1 : $result[0] =~ /^$self->{_token}$/))
	{
	    $result = qq/$result[0]$q/;
	}else{
	    $result = 
		$self->{po} . join($joiner => @result) . $self->{pc} .  $q;
	}
    }
    $self->{debug} and 
	print STDERR '<'x $self->{_indent}, " ", $result, "\n";
    $self->{_x} || $self->{debug} and $self->{_indent}--;

    $result;
}

sub _optim_cc{
    my $self = shift;
    @_ or return undef;
    if ($self->{quotemeta}){
	return  @_ ? @_ > 1  ? "[".join("",@_)."]" : $_[0] : undef;
    }
    # check '.'
    for (@_){
	$_ eq '.' and return '.';
    }
    my @char = @_;
    my ($positive, $negative) = ('','');
    my ($npos, $nneg) = (0, 0);
    for (@char){
	if    (s/^\[\^(.*)\]$/$1/){
	    $negative .= $_; $nneg += 2; next;
	}
	if (s/^\[(.*)\]$/$1/){
	    $positive .= $_; $npos += 2; next;
	}else{
	    #$positive .= length($_) eq 1 ? qq/\Q$_/ : $_;
	    $positive .= $_ eq '-' ? '\-' : $_;
	    $npos++;
	}
    }
    $nneg > 1 and $negative = qq/[^$negative]/;
    $npos > 1 and $positive = qq/[$positive]/;
    return $negative 
	?  $positive ? "(?:$positive|$negative)" : $negative
	    :  $positive;
}

1;
__END__

# Below is stub documentation for your module. You'd better edit it!

=head1 NAME

Regexp::List - builds regular expressions out of a list of words

=head1 SYNOPSIS

  use Regexp::List;
  my $l  = Regexp::List->new;
  my $re = $l->list2re(qw/foobar fooxar foozap fooza/);
  # $re is now qr/foo(?:[bx]ar|zap?)/

=head1 ABSTRACT

This module offers C<list2re> method that turns a list of words
into an optimized regular expression which matches all words therein.

The optimized regular expression is much more efficient than a
simple-minded '|'-concatenation thereof.

=head1 DESCRIPTION

This module use Object-Oriented approach so you can use this module as a base
and tweak its features.  This module is a base class of
L<Regexp::Optimizer>.

=head2 EXPORT

Since this is an OO module there is no symbol exported.

=head1 METHODS

This module offers methods below;

=over

=item $l = Regexp::List->new(I<< key=>value, ... >>)

Constructor.  When arguments are fed in I<< key => value, >> manner, it
sets attributes.  See C<< $l->set >> for details

=item $re = $l->list2re(I<list of words ...>)

Does the job.  Takes a list of words and turn it into an optimal
regular expresson.  See L</IMPLEMENTATION> to find out how it is
achieved.  If you want to know the underlying black magic even
further, see the source.

=item $l->set(I<< key => value, ... >>)

Sets attributes.  There are many attributes supported but let me
mention just a few that you may be interested.

=over

=item lookahead

Whether you prepend a lookahead assertion or not.  Default value is 1.
This module is smart enough to omit the assertion when you don't need one.

  $re = $l->list2re(qw/1 2 3 infinity/);
  # qr/(?=[123i])(?:[123]|infinity)/
  $re = $l->set(lookahead=>0)->list2re(qw/1 2 3 infinity/);
  # qr/(?:[123]|infinity)/

=item quotemeta

Whether you quote metacharacters or not.  Default is 1.  If you really
need this feature try L<Regexp::Optimizer> instead.

  @list = qw/3 3.14 3.14159265358979/;
  $re = $l->list2re(@list);
  # qr/3(?:\.14(?:159265358979)?)?)/
  $re = $l->set(lookahead=>0)->list2re(@list);
  # qr/3(?:.14(?:159265358979)?)?)/
  # which does match 3.14 but also "11+3=14"

=item modifies

Currently it accepts 'i', 'm', 's', and 'x', the same as regular
expression modifiers.

  @list = qw/Perl perl BASIC basic/;
  $re = $l->list2re(@list);
  # qr/(?=[BPbp])(?:[Pp]erl|BASIC|basic)/
  $re = $l->set(modifiers => 'i')->list2re(@list);
  # qr/(?=[bp])(?:perl|basic)/i
  print $l->set(modifiers => 'x')->list2re(@list);
  # Try for yourself;  Isn't itcute ?

=back

=item $l->expand($re);

Utility methods to expand a regular expression.  Handy when you
want to check the complex regexes.

=item $l->unexpand($re);

Utility methods to unexpand a regular expression.

=back

=head1 IMPLEMENTATION

This module optimizes the regular expression as follows.  Let's see
what happens when qw/foobar fooxar foozap fooza/ is fed

=over

=item trie building (common prefix aggregation)

first the corresponding I<trie> structure is built

       +- bar
  foo -+- xar
       +- za -+- p
              +- ''

=item common suffix aggregation

it checks if any leaf node can be optimized for common suffix.  In
this case we can do so to "bar" and "xar".

       +- b -+-ar
  foo -+- x -+
       +- za -+- p
              +- ''

=item character class conversion

If a branch contains more than two single characters, it turns it into
a character class.

  foo -+- [bx] --- ar
       +- za -+-p
              +- ''

=item empty leaf to C<?>

Empty leaf is converted to a '?' quantifier

  foo -+- [bx] --- ar
       +- za -+-p?


=item join all leafs into a group

And the final result is reached.

  foo(?:[bx]ar|zap?)

=back 

=head1 BENCHMARKS

This module is faily robust.  You can practically use this module to
find a regular expression that matches all words in a dictionary.
Here is a result by on perl 5.8.0, FreeBSD 4-Stable, Pentium III 800
Mhz with 512 MB RAM.

 # Sat May 31 09:11:06 2003 (     0.000000 s) Reading /usr/share/dict/words
 # Sat May 31 09:11:07 2003 (     0.847797 s) 235881 lines read.
 # Sat May 31 09:11:07 2003 (     0.000000 s) Making regexp.
 # Sat May 31 09:13:09 2003 (   121.596928 s) Done.
 # Sat May 31 09:13:09 2003 (     0.000000 s) Saving to t/words.rx
 # Sat May 31 09:13:09 2003 (     0.000000 s) Reading t/words.rx
 # Sat May 31 09:13:13 2003 (     3.679176 s) Done.
 # Sat May 31 09:13:13 2003 (     0.000000 s) Opening /usr/share/dict/words for comparison.
 # Sat May 31 09:13:13 2003 (     0.255222 s) /usr/share/dict/words:235881 lines found.
 # Sat May 31 09:13:13 2003 (     0.000000 s) Showtime!
 # 235881/235881
 # Sat May 31 10:44:17 2003 (  5464.370409 s) Done.
 # Sat May 31 10:44:17 2003 (  5464.370624 s) 43.167 matches/s

The result of optimization is obvious as the number of alteration
increases.  Here is a result of a benchmark which matches randomly
picked words against C</usr/share/dict/words>.

  ====        2 words
          Rate naive optim
  naive 1.79/s    --  -28%
  optim 2.49/s   39%    --

  ====      256 words
        s/iter naive optim
  naive   31.7    --  -81%
  optim   5.95  433%    --

=head1 SEE ALSO

L<Regexp::Optimizer> -- uses this module as its base

C<eg/> directory in this package contains example scripts.

=over

=item Perl standard documents

L<perltodo>, L<perlre>

=item CPAN Modules

L<Regexp::Presuf>, L<Text::Trie>

=item Books

Mastering Regular Expressions  L<http://www.oreilly.com/catalog/regex2/>

=back

=head1 AUTHOR

Dan Kogai <dankogai@dan.co.jp>, Kato Kazuyoshi <kato.kazuyoshi@gmail.com>

=head1 COPYRIGHT AND LICENSE

Copyright 2003 by Dan Kogai, All Rights Reserved.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself. 

=cut
